Tokenizer summary (#5467)
huggingface/transformers: docs/source/tokenizer_summary.rst

algorithms the pretrained model used. For instance, if we look at :class:`~transformers.BertTokenizer`, we can see it's
using :ref:`WordPiece <wordpiece>`.

Introduction to tokenization
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Splitting a text in smaller chunks is a task that's harder than it looks, and there are multiple ways of doing it. For
instance, let's look at the sentence "Don't you love 🤗 Transformers? We sure do." A first simple way of tokenizing
this text is just to split it by spaces, which would give:

::

    ["Don't", "you", "love", "🤗", "Transformers?", "We", "sure", "do."]

This is a nice first step, but if we look at the tokens "Transformers?" or "do.", we can see we can do better. Those
will be different than the tokens "Transformers" and "do" for our model, so we should probably take the punctuation
into account. This would give:

::

    ["Don", "'", "t", "you", "love", "🤗", "Transformers", "?", "We", "sure", "do", "."]

which is better already. One thing that is annoying though is how it dealt with "Don't". "Don't" stands for do not, so
it should probably be better tokenized as ``["Do", "n't"]``. This is where things start getting more complicated, and

> part of the reason each kind of model as its own tokenizer class. Depending on the rules we apply to split our texts

Reviewer: ```suggestion
part of the reason each kind of model has its own tokenizer class. Depending on the rules we apply to split our texts
```

into tokens, we'll get different tokenized versions of the same text. And of course, a given pretrained model won't
perform properly if you don't use the exact same rules as the persons who pretrained it.

`spaCy <https://spacy.io/>`__ and `Moses <http://www.statmt.org/moses/?n=Development.GetStarted>`__ are two popular
rule-based tokenizers. On the text above, they'd output something like:

::

    ["Do", "n't", "you", "love", "🤗", "Transformers", "?", "We", "sure", "do", "."]

Space/punctuation-tokenization and rule-based tokenization are both examples of word tokenization, which is splitting a
sentence into words. While it's the most intuitive way to separate texts in smaller chunks, it can have a problem when
you have a huge corpus: it usually yields a very big vocabulary. :doc:`Transformer XL <model_doc/transformerxl>` for

> instance uses space/punctuation-tokenization, and has a vocabulary size of 267,735!

Reviewer 2: In these lines as well as the following you often mention the word "vocabulary". As a newcomer, I'm not sure I would understand what that is.

Author: Good point, will try to add some definition.

> A huge vocabulary size means a huge embedding matrix, which will cause memory problem. TransformerXL deals with it by

Reviewer: ```suggestion
A huge vocabulary size means a huge embedding matrix, which will cause memory problems. TransformerXL deals with it by
```

vocabulary size greater than 50,000, especially if they are trained on a single language.

So if tokenizing on words is unsatisfactory, we could go on the opposite direction and simply tokenize on characters.
While it's very simple and would save a lot of memory, this doesn't allow the model to learn representations of texts
as meaningful as when using a word tokenization, leading to a loss of performance. So to get the best of both worlds,
all transformers models use a hybrid between word-level and character-level tokenization called subword tokenization.

Subword tokenization
^^^^^^^^^^^^^^^^^^^^

Subword tokenization algorithms rely on the principle that most common words should be left as is, but rare words
should be decomposed in meaningful subword units. For instance "annoyingly" might be considered a rare word and
decomposed as "annoying" and "ly". This is especially useful in agglutinative languages such as Turkish, where you can
form (almost) arbitrarily long complex words by stringing together some subwords.

This allows the model to keep a reasonable vocabulary while still learning useful representations for common words or

> subwords. This also give the ability to the model to process words it has never seen before, by decomposing them into

Reviewer: ```suggestion
subwords. This also gives the ability to the model to process words it has never seen before, by decomposing them into
```

> subwords. This also give the ability to the model to process words it has never seen before, by decomposing them into

Reviewer 2: ```suggestion
subwords. This also gives the ability to the model to process words it has never seen before, by decomposing them into

```

Another example is when we use the base :class:`~transformers.XLNetTokenizer` to tokenize our previous text:

::

    >>> from transformers import XLNetTokenizer
    >>> tokenizer = XLNetTokenizer.from_pretrained('xlnet-base-cased')
    >>> tokenizer.tokenize("Don't you love 🤗 Transformers? We sure do.")
    ['▁Don', "'", 't', '▁you', '▁love', '▁', '🤗', '▁', 'Transform', 'ers', '?', '▁We', '▁sure', '▁do', '.']

We'll get back to the meaning of those '▁' when we look at :ref:`SentencePiece <sentencepiece>` but you can see
Transformers has been split into "Transform" and "ers".

Let's now look at how the different subword tokenization algorithms work. Note that they all rely on some form of
training which is usually done on the corpus the corresponding model will be trained on.

.. _byte-pair-encoding:

Byte-Pair Encoding
~~~~~~~~~~~~~~~~~~

Byte-Pair Encoding was introduced in `this paper <https://arxiv.org/abs/1508.07909>`__. It relies on a pretokenizer
splitting the training data into words, which can be a simple space tokenization
(:doc:`GPT-2 <model_doc/gpt2>` and :doc:`Roberta <model_doc/roberta>` uses this for instance) or a rule-based tokenizer

> (:doc:`XLM <model_doc/xlm>` use Moses for most languages, as do :doc:`FlauBERT <model_doc/flaubert>`),

Reviewer 2: ```suggestion
(:doc:`XLM <model_doc/xlm>` use Moses for most languages, as does :doc:`FlauBERT <model_doc/flaubert>`),

```

:doc:`GPT <model_doc/gpt>` uses Spacy and ftfy) and, counts the frequency of each word in the training corpus.

> It then begins from the list of all characters, and will learn merge rules to form a new token from to symbols in the

Reviewer: ```suggestion
It then begins from the list of all characters, and will learn merge rules to form a new token from two symbols in the
```

> It then begins from the list of all characters, and will learn merge rules to form a new token from to symbols in the

Reviewer 2: ```suggestion
It then begins from the list of all characters, and will learn merge rules to form a new token from two symbols in the

```

Then the base vocabulary is ['b', 'g', 'h', 'n', 'p', 's', 'u'] and all our words are first split by character:

::

    ('h' 'u' 'g', 10), ('p' 'u' 'g', 5), ('p' 'u' 'n', 12), ('b' 'u' 'n', 4), ('h' 'u' 'g' 's', 5)

We then take each pair of symbols and look at the most frequent. For instance 'hu' is present `10 + 5 = 15` times (10
times in the 10 occurrences of 'hug', 5 times in the 5 occurrences of 'hugs'). The most frequent here is 'ug', present
`10 + 5 + 2 + 5 = 22` times in total. So the first merge rule the tokenizer learns is to group all 'u' and 'g' together
then it adds 'ug' to the vocabulary. Our corpus then becomes

::

    ('h' 'ug', 10), ('p' 'ug', 5), ('p' 'u' 'n', 12), ('b' 'u' 'n', 4), ('h' 'ug' 's', 5)

and we continue by looking at the next most common pair of symbols. It's 'un', present 16 times, so we merge those two
and add 'un' to the vocabulary. Then it's 'hug' (as 'h' + 'ug'), present 15 times, so we merge those two and add 'hug'
to the vocabulary.

At this stage, the vocabulary is ``['b', 'g', 'h', 'n', 'p', 's', 'u', 'ug', 'un', 'hug']`` and our corpus is
represented as

::

    ('hug', 10), ('p' 'ug', 5), ('p' 'un', 12), ('b' 'un', 4), ('hug' 's', 5)

> If we stop there, the tokenizer can apply the rules it learned to new words (as long as they don't have characters that

Reviewer: ```suggestion
If we stop there, the tokenizer can apply the rules it learned to new words (as long as they don't contain characters that
```

> are in the base vocabulary). For instance 'bug' would be tokenized as ``['b', 'ug']`` but mug would be tokenized as

Reviewer: ```suggestion
were not in the base vocabulary). For instance 'bug' would be tokenized as ``['b', 'ug']`` but mug would be tokenized as
```

``['<unk>', 'ug']`` since the 'm' is not in the base vocabulary. This doesn't happen to letters in general (since the

> base corpus uses all of them), but to special characters like emoticons.

Reviewer: ```suggestion
base corpus uses all of them), but to special characters like emojis.
```

> base corpus uses all of them), but to special characters like emoticons.

Reviewer 2: ```suggestion
base corpus uses all of them), but to special characters like emojis.

```
